api/task_manager.cc

```diff
@@ -180,7 +181,51 @@ void set_task_manager(http_context& ctx, routes& r) {
                 return make_foreign(task);
             });
         }));
-        co_return co_await retrieve_status(std::move(task));
+        auto s = co_await retrieve_status(task);
+        co_return make_status(s);
+    });
+
+    tm::get_task_status_recursively.set(r, [&ctx] (std::unique_ptr<request> req) -> future<json::json_return_type> {
+        using chunked_statuses = utils::chunked_vector<full_task_status>;
+
+        auto& _ctx = ctx;
+        auto id = tasks::task_id{utils::UUID{req->param["task_id"]}};
+        chunked_statuses res;
+
+        // Push the requested task's status.
+        auto task = co_await tasks::task_manager::invoke_on_task(_ctx.tm, id, std::function([] (tasks::task_manager::task_ptr task) -> future<tasks::task_manager::foreign_task_ptr> {
+            auto state = task->get_status().state;
+            if (state == tasks::task_manager::task_state::done || state == tasks::task_manager::task_state::failed) {
+                task->unregister_task();
+            }
+            co_return std::move(task);
+        }));
+        res.push_back(co_await retrieve_status(task));
+
+        // Push children's statuses in BFS order.
+        size_t current = 0;
+        while (current < res.size()) {
+            auto& current_status = res[current];
+
+            for (auto& t : current_status.child_tasks) {
+                res.push_back(co_await retrieve_status(t));
```

Comment:
Since this function yields, it should rely on this vector being valid.
Instead, like I suggested earlier, the task status should keep the children ids and they should be looked up here.

Reply:
I extensively focused on minimizing the number of accesses to different shards while unreasonably ignoring all other aspects.

Recursion is done with `current` serving a queue for BFS. Now it would not be so streightforward though.

I can see two approches:

1. In `full_task_status` we keep a vector of task_info (task_id + shard) and look for each task in the module. This is rather problematic, because we could have already removed some of child tasks from module so we won't give user the complete information.

2. We can access information about all descendants even if they were removed from module but we need to simulate a queue somehow (for bfs). To do so we can:
- copy all foreign_ptrs in an actual queue (which means an access to owner shard for each pointer)
- keep a vector of reference wrappers to vectors of children (something similar to what I tried to achieve in first version). All of them will be alive while processed (since the parent will be) but this sounds pretty obscure.

Reply:
Which one do you find better?